### PR TITLE
[ADD] mrp_project

### PR DESCRIPTION
**Pull request by pedrobaeza**
# 
# MRP Project Link

It links projects and tasks to _manufacturing orders_ (MO) and
_work orders_ (WO).
# Usage

In a manufacturing order (MO), you can select a project to be attached to it.
If none is selected, when the MO is confirmed, a project is automatically
created.

When the MO starts, a task is created and assigned to the order.

### CODE REVIEW

**Comment by max3903** [comment]
@pedrobaeza During my test, I ended up with one task for the manufacturing order and one task for each work order. Is it intentional ? What is the requirement behind this behavior ? I was expecting tasks only for the work orders.

Otherwise :+1:

**Comment by pedrobaeza** [comment]
This is intentional to provide one general task for manufacturing order where placing works (times) for manufacturing pre-operations, post-operations... and another per work order for having granularity for each of the phases of the production. In a later module that I'm proposing later today, it can even be duplicated and assigned to each of the workers assigned on the workcenter.

**Comment by anajuaristi** [comment]
If this could help, I will try to explain the functional behavior of this desing.

Apart the one explained by @pedrobaeza, there is some companies where is too complex to assing a task for each operator that is working on MO.  They even don't know who is going to work in one operation depending on the production on each moment.
For this kind of company is not needed to create a task for each operator. They can just go to general MO task and report made work directly on it.
So, If you set 0 workers on routing line, system will not create additional tasks for workers.

But... if you set workers more than cero on routing line system will create as many tasks as workers you indicated.
Additionally, if you set the list of workers on "machine/workcenter" (you need to activate the option on manufacturing settings), system will assing the workers on tasks.
If the number of workers is the same than the list of people assigned to the machine everything is right
If the number of workers is bigger than the list, there will be some "unassigned" tasks
if the number of workers is smaller than the list, system will assing workers following the sequence indicated on list.

@pedrobaeza I don't know if maybe it would be interesting adding this kind of explanation in readme.

Wishing this helps... 
Ana

**Comment by max3903** [comment]
@pedrobaeza @anajuaristi Ok, I understand. Maybe you could add in the description of this task something like:

```
Meta task of the MO to enter time not related to any work orders, 
like operator assignment, preparation, cleaning, etc.
```

**Comment by rafaelbn** [comment]
>  So, If you set 0 workers on routing line, system will not create additional tasks for workers.
> 
> But... if you set workers more than cero on routing line system will create as many tasks as workers you indicated.
> 
>  Additionally, if you set the list of workers on "machine/workcenter" (you need to activate the option on manufacturing settings), system will assing the workers on tasks.
- @anajuaristi I don't see any of this options you say
- This modules links mrp with project. We can see from tasks the work order related, where from work order we can see the task related?

**Comment by pedrobaeza** [comment]
What @anajuaristi says is going to be in _mrp_operations_project_ module.

**Comment by rafaelbn** [comment]
> This is intentional to provide one general task for manufacturing order where placing works (times) for manufacturing pre-operations, post-operations... 

Then please add in README configuration section that you must Go to Settings - configuration - project and enable Log work activities on tasks . If you don't enable this doesnt have any sense.

Will be Any link from Work Orders to task?

**Comment by anajuaristi** [comment]
From task to mo and wo
From mo to project and tasks

We did no include any link from wo to tasks because it's suposed that
workers will go directly to tasks. Not necesary to go through operations

Takss can be filtered and grouped by mo, wo and of course by user

It would be really easy to add shortcut from wo to tasks if needed

**Comment by rafaelbn** [comment]
OK, then:
- Redirecting to workers to input time in task make you need to add MRP fields to tasks
- This is not the best UX way. You have already WO, why you don't make that this workers input time in the WO?
- Will be clearest and faster for worker to input times and assign work in WO view, and further more you will not have to give access to Project to the workers.
- Odoo thought WO view simple an easy to them, Task view is more complicated for this kind of workers.
- If you would not redirect to workers to tasks and let input time in WO then you can have a clearest view form in task.

I would like to here more opinions in this matter. IMHO this must be done by another module, let's say **mrp_project_task** and leave this **mrp_project** the most lighter and cleaner without MRP fields in tasks

**Comment by pedrobaeza** [comment]
Please, don't continue here this discussion because this module is not the proper one. You need to see _mrp_operations_project_ where the link is made.

**Comment by rafaelbn** [comment]
I'm talking about this module:
- Please add filter is from Manufacturing to avoid this


- Please add a tab TASK in MO as added in Task a tab MRP data (<- this both things is what I said I would be better in mrp_project_task)

**Comment by pedrobaeza** [comment]
Tasks doesn't make sense in the MO. They're added on the WO. A link to the project in the MO does, what is there. You click, and navigate to the project then.

**Comment by JordiBForgeFlow** [comment]
Hi I already commented to Ana before. I do not really see the benefit of that. If it is just to enter time you can enter timesheet lines associated to WO.

You don't manage projects and tasks because it's a convenient place to enter time.. The world does not run like this :S

**Comment by rafaelbn** [comment]
@jbeficent about you comment, I could say: this module do this to organize work, assign task, schedule, plan... what would you answer me about this?

**Comment by JordiBForgeFlow** [comment]
This is an alternative:
https://github.com/Eficent/eficent-odoo-addons/tree/7.0/mrp_workorder_line_work

https://github.com/Eficent/eficent-odoo-addons/tree/7.0/mrp_workorder_line_work_timesheet

**Comment by anajuaristi** [comment]
World runs depending on glasses you use to take a look

Please answer this question:
Supose you have got several workers to register time in one WO and you need
to assing an stimated time to each of them. You would need to heavily
modify wo to include that behavior.
If you use tasks... you are not reinventing the weal. All people on company
register their time same way. There is a single entry point of data.
You can even control engineers and designers dedicated time for the mo/wo.
This way is also easier to extract unified reports.
About "undefined" group by is also correct since you can easily distinguiss
registers coming from Mo or Wo and the ones that are coming from any other
reason.
On the othe side the only single object you need to allow workers to
include their time is timesheet table on task. You could even avoid then to
go into the task defining a new menu entry. It's quite simple.

To finish... if someone need also having timesheet directly on WO both ways
are compatible. But even WO form screen is quite complex form depending on
profile of the workers so, selecting this option should not be simplicity
but any other reasons

File: mrp_project/README.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+================
+MRP Project Link
+================
+
+It links projects and tasks to *manufacturing orders* (MO) and
```
Comment (elicoidal) [inline mrp_project/README.rst:9]: s/it/This module

File: mrp_project/README.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+*work orders* (WO).
+
+Usage
+=====
+
+In a manufacturing order (MO), you can select a project to be attached to it.
+If none is selected, when the MO is confirmed, a project is automatically
```
Comment (elicoidal) [inline mrp_project/README.rst:16]: If none is selected, a project is automatically created when the MO is confirmed.

**Comment by rafaelbn** [comment]
> World runs depending on glasses you use to take a look

I agree, this module must be **more general purpose** and not only one type of glasses, that why we must split it it two modules:

## Granularity
- mrp_project (better mrp_analytic) only links MO with account.analytic.account as Odoo makes in projects (see screenshot 1)
- mrp_project (better mrp_project_task) add the functionality of creating tasks from MO and adding MRP data in task

###### Screenshoot 1



> Please answer this question: Suppose you have got several workers to register time in one WO and you need to assign an estimated time to each of them. You would need to heavily modify WO to include that behavior.

@anajuaristi , you mix future functionalities in this module. Please draw the design as for example with "Training module" as @yajo made in https://github.com/OCA/event/pull/14#issuecomment-159230753 with a doc https://docs.google.com/document/d/1WiOB1yAWTR2aROpM6O632QWXwgG4hLRUpTHU2Pv4yEQ/edit#heading=h.vjtptnyqc313 and a diagram like @dreispt made https://drive.google.com/file/d/0B72Vums-E7oodlplWlhMQ2MxNVk/view?usp=sharing
- If I suppose that I have several workers to register time in one WO and I need to assign an time to each of them, then 2 options:
  - I will need to create a task for each worker **as you say** but not in this module, in future new modules.

Now I can say, please Ana, suppose you need yo link a MO with an account analytic account but you **don't need to assign a time each of the workers, you only want to know the total time all the workers spent in the WO**, and you don't want any task created, only input time.

As @jbeficent say https://github.com/OCA/manufacture/pull/56#issuecomment-159999514 the are more points of view and we must give granularity for giving more value to OCA modules

## Analysis, Cost analysis

> @max3903 I ended up with one task for the manufacturing order and one task for each work order. Is it intentional ? What is the requirement behind this behavior ?
> 
> @pedrobaeza This is intentional to provide one general task for manufacturing order where placing works (times) 

If you want to place works means input times, if you want to input times is because you want to analyze whats going on with this time. Conclusion: account.analytic.account., not project. In fact, **every project has account.analytic.account for this purpose.

The same makes Odoo see above screenshot 1

> @pedrobaeza In a later module that I'm proposing later today

I think we must draw a most general view of the puzzle. If today we only think one case of use: **create task from MO** TOMORROW this will fight with **creating MO from task** with is one of the objetive for Odoo to win SAP mature standards (in SAP project task and work orders is the same table, SAP lets manage MRP from Project using BoM without MO).

## Conclusion

Please think in all possibilities, this module must be core and NOT determinate future modules and much less cost analysis.

**Comment by JordiBForgeFlow** [comment]
The link of MO to analytic account must be a separate module. It must be the cornerstone to link MO costs, independently how. 

See:
https://github.com/Eficent/eficent-odoo-addons/tree/7.0/mrp_analytic

**Comment by anajuaristi** [comment]
The behavior of the link with projects instead of analytic acccount(that is
there but invisible way) is the core problem where project creates an
analytic account but if you first create analytic account you can not
create the project afterwards to link with the previously created analytic
account.
The steps always have to be first project, after analytic account otherwise
there is the risk of duplicating analytic account.
Be carefull with this point because it generates heavy issues so if we need
to vote about changing project to analytic account I would say NO.

On the other side, if you need just to register global amount of cost of
all the workers... you only need to set 0 workers on routing stepts and no
task for workers will be created. You will have only the MO project +
analytic account so you can directly put the analytic like using the form
you want.

If you technicaly want to separate modules on several ones, including
dependencies, I agree but by the moment I don't see any functional reason
to change functionality on them.

2015-11-27 9:12 GMT+01:00 Jordi Ballester Alomar notifications@github.com:

> The link of MO to analytic account must be a separate module. It must be
> the cornerstone to link MO costs, independently how.
> 
> See:
> https://github.com/Eficent/eficent-odoo-addons/tree/7.0/mrp_analytic
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/manufacture/pull/56#issuecomment-160065290.

## 

CEO Avanzosc, S.L http://www.avanzosc.es : Office phone / Tfono oficina:
(+34) 943 02 69 02
Ana Juaristi Olalde http://www.anajuaristi.com/: Personal phone: 677 93
42 59. User/usuario skype: Avanzosc
www.openerpsite.com

_El contenido de esta comunicación y de toda su documentación anexa es
confidencial y se dirige exclusivamente a su destinatario. El uso no
autorizado de esta información está prohibido por la legislación vigente.
Si usted no es el destinatario le rogamos nos lo indique, no comunique su
contenido a terceros y proceda a su destrucción. Disculpe las molestias que
le haya ocasionado la recepción indebida de este e-mail. Sus datos figuran
en un fichero cuyo titular es Avanzosc, S.L., a quien usted puede dirigirse
para ejercer sus derechos de acceso, rectificación, cancelación y oposición
en Klara Donea 13, 20720, Azkoitia (Gipuzkoa), Tef. 943 02 69 02 -
*_administracion@avanzosc.
soporte@avanzosc.eses*

_Komunikazio honen edukia eta dokumentazio erantsia konfidentziala da eta
hartzaileak bakarrik jaso beharko luke. Indarrean dagoen legeriak debekatu
egiten du bertan eskainitako informazioa baimenik gabe erabiltzea.
Komunikazioa zuri iritsi bazaizu, baina zu ez bazara hartzailea, mesedez,
guri jakinarazi, eta jasotako informazioa ez inori jakinarazi eta suntsitu.
Barkatu okerreko email hau jasotzeak eragindako eragozpenak. Zure datuak
Avanzosc, S.L. enpresaren fitxategietan sartuta daude. Zure datuak atzitzea
eska dezakezu, bai eta, datuak zuzentzea, ezereztea eta tratamenduari aurka
egitea ere. Horretarako, enpresara jo dezakezu, helbide honetan: *_Klara
Donea 13*\* 20720, Azkoitia (Gipuzkoa), telefonoa: 943 02 69 02 - *
_administracion@avanzosc. soporte@avanzosc.eses_ _This message and all
documents attached to it are confidential and intended only for the person
or entity to which it is addressed. Any use of this information by
unauthorised persons is prohibited under current legislation. If you
received this message by error, please advise us, destroy it and refrain
from communicating its contents to third parties. We apologise for any
inconvenience receiving this email improperly may cause to you. Your
personal data are included in a file owned by Avanzosc, S.L. If you want to
exercise your rights of access, correction, erasure and objection you can
contact the Controller at *_Klara Donea 13*\* 20720, Azkoitia (Gipuzkoa), T:
943 02 69 02 – administracion@avanzosc. soporte@avanzosc.eses*

**Comment by JordiBForgeFlow** [comment]
But Ana you CAN create a project from an existing analytoc account. Press the 'Task' flag.

**Comment by anajuaristi** [comment]
Ok, I see.

This functionality has been included on version 8. If you create an
analytic account and check task on the form, it creates also the project
automatically. It's good!
Thank you Jordi because I didn't know about this.

So... OK, this way I don't mind if MO is showing analytic account or
project since we can create the account also as project.

Who is going to start making this "cleaning" and "module spliting" tasks?

2015-11-27 9:28 GMT+01:00 Jordi Ballester Alomar notifications@github.com:

> But Ana you CAN create a project from an existing analytoc account. Press
> the 'Task' flag.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/manufacture/pull/56#issuecomment-160067681.

## 

CEO Avanzosc, S.L http://www.avanzosc.es : Office phone / Tfono oficina:
(+34) 943 02 69 02
Ana Juaristi Olalde http://www.anajuaristi.com/: Personal phone: 677 93
42 59. User/usuario skype: Avanzosc
www.openerpsite.com

_El contenido de esta comunicación y de toda su documentación anexa es
confidencial y se dirige exclusivamente a su destinatario. El uso no
autorizado de esta información está prohibido por la legislación vigente.
Si usted no es el destinatario le rogamos nos lo indique, no comunique su
contenido a terceros y proceda a su destrucción. Disculpe las molestias que
le haya ocasionado la recepción indebida de este e-mail. Sus datos figuran
en un fichero cuyo titular es Avanzosc, S.L., a quien usted puede dirigirse
para ejercer sus derechos de acceso, rectificación, cancelación y oposición
en Klara Donea 13, 20720, Azkoitia (Gipuzkoa), Tef. 943 02 69 02 -
*_administracion@avanzosc.
soporte@avanzosc.eses*

_Komunikazio honen edukia eta dokumentazio erantsia konfidentziala da eta
hartzaileak bakarrik jaso beharko luke. Indarrean dagoen legeriak debekatu
egiten du bertan eskainitako informazioa baimenik gabe erabiltzea.
Komunikazioa zuri iritsi bazaizu, baina zu ez bazara hartzailea, mesedez,
guri jakinarazi, eta jasotako informazioa ez inori jakinarazi eta suntsitu.
Barkatu okerreko email hau jasotzeak eragindako eragozpenak. Zure datuak
Avanzosc, S.L. enpresaren fitxategietan sartuta daude. Zure datuak atzitzea
eska dezakezu, bai eta, datuak zuzentzea, ezereztea eta tratamenduari aurka
egitea ere. Horretarako, enpresara jo dezakezu, helbide honetan: *_Klara
Donea 13*\* 20720, Azkoitia (Gipuzkoa), telefonoa: 943 02 69 02 - *
_administracion@avanzosc. soporte@avanzosc.eses_ _This message and all
documents attached to it are confidential and intended only for the person
or entity to which it is addressed. Any use of this information by
unauthorised persons is prohibited under current legislation. If you
received this message by error, please advise us, destroy it and refrain
from communicating its contents to third parties. We apologise for any
inconvenience receiving this email improperly may cause to you. Your
personal data are included in a file owned by Avanzosc, S.L. If you want to
exercise your rights of access, correction, erasure and objection you can
contact the Controller at *_Klara Donea 13*\* 20720, Azkoitia (Gipuzkoa), T:
943 02 69 02 – administracion@avanzosc. soporte@avanzosc.eses*

**Comment by rafaelbn** [comment]
> The behavior of the link with projects instead of analytic acccount(that is there but invisible way) is the core problem where project creates an analytic account but if you first create analytic account you can not create the project afterwards to link with the previously created analytic account. The steps always have to be first project, after analytic account otherwise there is the risk of duplicating analytic account
- This is not as you say, you can create a project from an analytic account without problem

> The behavior of the link with projects instead of analytic acccount (that is there but invisible way)
- This is not valid, this must be separated in 2 modules, invisible way must be 2 visible separated modules

**Comment by rafaelbn** [comment]
> Who is going to start making this "cleaning" and "module splitting" tasks?

OK, Antiun assume it.

Objective:
- mrp_analytic: relation MO with account.analytic.account
- mrp_project: relation MO with project.project (if mrp.analytic installed then is the same)
- mrp_project_task: creates task from MO and add in task add tab mrp data

@jbeficent @anajuaristi @max3903 @pedrobaeza  are you agree?

**Comment by anajuaristi** [comment]
Yes. I agree.

At the end, we are using analytic account on MO to generate stimated and
real cost, so if you change names or split functionality, we only need to
change the dependencies on that modules so changes are not heavily
impacting it's not so much work to change that.

@pedrobaeza opinion?

2015-11-27 9:52 GMT+01:00 Rafael Blasco notifications@github.com:

> Who is going to start making this "cleaning" and "module splitting" tasks?
> 
> OK, Antiun assume it.
> 
> Objective:
> - mrp_analytic: relation MO with account.analytic.account
> - mrp_project: relation MO with project.project (if mrp.analytic
>   installed then is the same)
> - mrp_project_task: creates task from MO and add in task add tab mrp
>   data
> 
> @jbeficent https://github.com/jbeficent @anajuaristi
> https://github.com/anajuaristi @max3903 https://github.com/max3903
> @pedrobaeza https://github.com/pedrobaeza are you agree?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/manufacture/pull/56#issuecomment-160075262.

## 

CEO Avanzosc, S.L http://www.avanzosc.es : Office phone / Tfono oficina:
(+34) 943 02 69 02
Ana Juaristi Olalde http://www.anajuaristi.com/: Personal phone: 677 93
42 59. User/usuario skype: Avanzosc
www.openerpsite.com

_El contenido de esta comunicación y de toda su documentación anexa es
confidencial y se dirige exclusivamente a su destinatario. El uso no
autorizado de esta información está prohibido por la legislación vigente.
Si usted no es el destinatario le rogamos nos lo indique, no comunique su
contenido a terceros y proceda a su destrucción. Disculpe las molestias que
le haya ocasionado la recepción indebida de este e-mail. Sus datos figuran
en un fichero cuyo titular es Avanzosc, S.L., a quien usted puede dirigirse
para ejercer sus derechos de acceso, rectificación, cancelación y oposición
en Klara Donea 13, 20720, Azkoitia (Gipuzkoa), Tef. 943 02 69 02 -
*_administracion@avanzosc.
soporte@avanzosc.eses*

_Komunikazio honen edukia eta dokumentazio erantsia konfidentziala da eta
hartzaileak bakarrik jaso beharko luke. Indarrean dagoen legeriak debekatu
egiten du bertan eskainitako informazioa baimenik gabe erabiltzea.
Komunikazioa zuri iritsi bazaizu, baina zu ez bazara hartzailea, mesedez,
guri jakinarazi, eta jasotako informazioa ez inori jakinarazi eta suntsitu.
Barkatu okerreko email hau jasotzeak eragindako eragozpenak. Zure datuak
Avanzosc, S.L. enpresaren fitxategietan sartuta daude. Zure datuak atzitzea
eska dezakezu, bai eta, datuak zuzentzea, ezereztea eta tratamenduari aurka
egitea ere. Horretarako, enpresara jo dezakezu, helbide honetan: *_Klara
Donea 13*\* 20720, Azkoitia (Gipuzkoa), telefonoa: 943 02 69 02 - *
_administracion@avanzosc. soporte@avanzosc.eses_ _This message and all
documents attached to it are confidential and intended only for the person
or entity to which it is addressed. Any use of this information by
unauthorised persons is prohibited under current legislation. If you
received this message by error, please advise us, destroy it and refrain
from communicating its contents to third parties. We apologise for any
inconvenience receiving this email improperly may cause to you. Your
personal data are included in a file owned by Avanzosc, S.L. If you want to
exercise your rights of access, correction, erasure and objection you can
contact the Controller at *_Klara Donea 13*\* 20720, Azkoitia (Gipuzkoa), T:
943 02 69 02 – administracion@avanzosc. soporte@avanzosc.eses*

**Comment by pedrobaeza** [comment]
OK, I will do it.

**Comment by JordiBForgeFlow** [comment]
@pedrobaeza I have created a new PR for module mrp_production_order_analytic: https://github.com/OCA/manufacture/pull/58

The name mrp_analytic is way too broad concept IMHO. We need to be specific in the naming of what this module does.

There are many modules provided by other authors such as Vauxoo, OdooMRP, Eficent,... that make us of the analytic account in manufacturing orders, but using different approaches. These should be incorporated, as separate modules.

For example, the concept of copying the analytic account from the MO to the associated stock moves is essential for us, but others may think that it is not.. so we can keep it as a separate module.

**Comment by pedrobaeza** [comment]
Why? I said yesterday that I was going to do it. We have double efforts. For what are we coordinating here?

Anyway, the name _mrp_analytic_ is the right one, because the rest of the modules are not OCA, so this one should be the reference one in one hand, and can group all the stuff related to link analytic with mrp in the other, although now only covers production orders.

Another topic is the repository: all the transversal modules to analytic are in account-analytic, so we should host this module in that repo.

**Comment by JordiBForgeFlow** [comment]
Ok, then disregard that one
El dia 28/11/2015 10:45, "Pedro M. Baeza" notifications@github.com va
escriure:

**Comment by pedrobaeza** [comment]
@rafaelbn, @anajuaristi, what do you think about the target for mrp_analytic: here of in OCA/account-analytic?

**Comment by anajuaristi** [comment]
Both would be correct.
Maybe people lookign for it would go first to account_analytics. I don't know. Let's listen to others opinion

**Comment by pedrobaeza** [comment]
I'm opening a new PR to clean comment history.

**Pull request closed without merging**